glean-core/android/src/main/java/mozilla/telemetry/glean/private/DatetimeMetricType.kt

```diff
@@ -19,19 +23,47 @@ import java.util.Date
  * Instances of this class type are automatically generated by the parsers at build time,
  * allowing developers to record values that were previously registered in the metrics.yaml file.
  */
-class DatetimeMetricType(
-    disabled: Boolean,
-    category: String,
-    lifetime: Lifetime,
-    name: String,
-    val sendInPings: List<String>,
-    timeUnit: TimeUnit = TimeUnit.Minute
+class DatetimeMetricType internal constructor(
+    var handle: Long,
+    val sendInPings: List<String>
 ) {
+    /**
+     * The public constructor used by automatically generated metrics.
+     */
+    constructor(
+        disabled: Boolean,
+        category: String,
+        lifetime: Lifetime,
+        name: String,
+        sendInPings: List<String>,
+        timeUnit: TimeUnit = TimeUnit.Minute
+    ) : this(handle = 0, sendInPings = sendInPings) {
+        val ffiPingsList = StringArray(sendInPings.toTypedArray(), "utf-8")
+        this.handle = LibGleanFFI.INSTANCE.glean_new_datetime_metric(
+            category = category,
+            name = name,
+            send_in_pings = ffiPingsList,
+            send_in_pings_len = sendInPings.size,
+            lifetime = lifetime.ordinal,
+            disabled = disabled.toByte(),
+            time_unit = timeUnit.ordinal
+        )
+    }
+
+    protected fun finalize() {
+        if (this.handle != 0L) {
+            val error = RustError.ByReference()
+            LibGleanFFI.INSTANCE.glean_destroy_datetime_metric(this.handle, error)
+        }
+    }
 
-    // private val logger = Logger("glean/DatetimeMetricType")
+    fun shouldRecord(): Boolean {
```

Comment:
I guess this should probably be `private`, also.  (Though this pattern is throughout the code -- file a separate bug if that makes more sense).

Reply:
Yes, we should refactor this. I filed [bug 1556650](https://bugzilla.mozilla.org/show_bug.cgi?id=1556650) for it.